stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -95,46 +95,53 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType, M,
     size_t tuple_size_ = sizeof...(Args);
 
     double sum_;
+    double* sliced_partials_;
     Eigen::VectorXd args_adjoints_;
 
-    recursive_reducer(size_t num_shared_terms, const std::vector<M>& vmapped,
-                      const Args&... args)
+    recursive_reducer(size_t num_shared_terms, double* sliced_partials,
+                      const std::vector<M>& vmapped, const Args&... args)
         : num_shared_terms_(num_shared_terms),
           vmapped_(vmapped),
           args_tuple_(args...),
           sum_(0.0),
+          sliced_partials_(sliced_partials),
           args_adjoints_(Eigen::VectorXd::Zero(num_shared_terms_)) {}
 
     recursive_reducer(recursive_reducer& other, tbb::split)
         : num_shared_terms_(other.num_shared_terms_),
           vmapped_(other.vmapped_),
           args_tuple_(other.args_tuple_),
           sum_(0.0),
+          sliced_partials_(other.sliced_partials_),
           args_adjoints_(Eigen::VectorXd::Zero(num_shared_terms_)) {}
 
     void operator()(const tbb::blocked_range<size_t>& r) {
       if (r.empty())
         return;
 
-      auto start = vmapped_.begin();
-      std::advance(start, r.begin());
-      auto end = vmapped_.begin();
-      std::advance(end, r.end());
-      std::vector<M> sub_slice(start, end);
+      std::vector<M> sub_slice;
+      sub_slice.reserve(r.end() - r.begin());
```

Comment:
@bbbales2 this code runs in any of the tbb threads. As we should really not leave anything behind on the child thread AD tapes, we must not create anything with `var` outside of a nested AD block. So this really has to go into the nested block as it will be left behind on the child threads AD tape, since its never cleaned away.

As a side note: you can call `r.size()` which is easier to read.

Reply:
I had a version that satisfies these requirements:

```
template <typename T>                                                                                                                                                                                                                                                       
std::vector<T> deep_copy_slice(const std::vector<T>& vmapped_, size_t begin, size_t end) {                                                                                                                                                                                            
  std::vector<T> sub_slice;                                                                                                                                                                                                                                                           
  sub_slice.reserve(end - begin);                                                                                                                                                                                                                                                     
  for (int i = begin; i < end; ++i) {                                                                                                                                                                                                                                                 
    sub_slice.emplace_back(vmapped_[i]);                                                                                                                                                                                                                                              
  }                                                                                                                                                                                                                                                                                   
  return sub_slice;                                                                                                                                                                                                                                                                   
}                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                          
std::vector<var> deep_copy_slice(const std::vector<var>& vmapped_, size_t begin, size_t end) {                                                                                                                                                                                        
  std::vector<var> sub_slice;                                                                                                                                                                                                                                                         
  sub_slice.reserve(end - begin);                                                                                                                                                                                                                                                     
  for (int i = begin; i < end; ++i) {                                                                                                                                                                                                                                                 
    sub_slice.emplace_back(vmapped_[i].val());                                                                                                                                                                                                                                        
  }                                                                                                                                                                                                                                                                                   
  return sub_slice;                                                                                                                                                                                                                                                                   
}                                                                                                                                                                                                                                                                                     
                                                                                                                                                                                                                                                                                          
void operator()(const tbb::blocked_range<size_t>& r) {                                                                                                                                                                                                                                
  if (r.empty())                                                                                                                                                                                                                                                                      
    return;                                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                                          
  try {                                                                                                                                                                                                                                                                               
    start_nested();                                                                                                                                                                                                                                                                   
                                                                                                                                                                                                                                                                                          
   // create a deep copy of all var's so that these are not                                                                                                                                                                                                                          
   // linked to any outer AD tree                                                                                                                                                                                                                                                    
   auto local_sub_slice = deep_copy_slice(vmapped_, r.begin(), r.end());                                                                                                                                                                                                             
...
```

Reply:
Oh I guess you could reduce that to one function with ```value_of```.

Reply:
the `value_of` which we have is not good enough to make a true `value_of` something which has a lot of structure in it.

Have a look at what I just put in here.

The version you had before looks also neat (I did not see it before I hacked up mine - so I have no opinion on whats better).